Filemanager: insert TitleBar

hius07 opened the pull request:
![01](https://user-images.githubusercontent.com/62179190/148971965-7e154c32-5059-4902-91f1-36617c157d8d.png)

![02](https://user-images.githubusercontent.com/62179190/148971995-d808e36d-678b-4229-8e39-4915166db93d.png)

![03](https://user-images.githubusercontent.com/62179190/148971998-656733ba-fec9-42c5-b957-08dc953783a4.png)

![04](https://user-images.githubusercontent.com/62179190/148971999-67dea3c2-9322-4d8f-986d-b25f1beb7f9d.png)

DPI 480
![05](https://user-images.githubusercontent.com/62179190/148972006-be24be2d-eaa8-4959-8907-effc17c3eaf1.png)

Compare titles in FM and in Menu widget

![06](https://user-images.githubusercontent.com/62179190/148972010-5147442a-3d3e-45aa-a229-e835dfada6d1.png)

![07](https://user-images.githubusercontent.com/62179190/148972013-bf18161a-d71a-4a45-879a-4f7914b92eae.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/8648)
<!-- Reviewable:end -->

hius07 commented:
Not done: a gesture assigned to top-right corner long-press doesn't work. IconButton eats the long-press.
Can we add a property to be passed through TitleBar to IconButton to force propagation?
Or is there a better way?

poire-z commented:
Might try to make these default values `nil` (not sure why I made them no-op functions):
https://github.com/koreader/koreader/blob/1bfb49c6a586a984fc99515d2fa69b8da125f75a/frontend/ui/widget/titlebar.lua#L50-L51
https://github.com/koreader/koreader/blob/1bfb49c6a586a984fc99515d2fa69b8da125f75a/frontend/ui/widget/titlebar.lua#L56-L57
as then, IconButton should let the tap propagate:
https://github.com/koreader/koreader/blob/1bfb49c6a586a984fc99515d2fa69b8da125f75a/frontend/ui/widget/iconbutton.lua#L103-L104
And don't provide the callback to your TitleBar, so the default `nil` is used.

hius07 commented:
That works, but, for example, close button in History propagates long-press to the below file manager.

hius07 commented:
If hold_callback is nil, a flag to allow/deny propagation is needed.

hius07 commented:
Interesting to see incremental vertical shift (rounding when calculating item height?).
Needs investigation.

![08](https://user-images.githubusercontent.com/62179190/148984066-5dedbaf3-6a61-4153-a4f4-22eec0deb3fc.png)

poire-z commented:
> If hold_callback is nil, a flag to allow/deny propagation is needed.

Then, keep the dummy functions as defaults, and provide `right_icon_tap/hold_callback = false` when they should propagate (in most other TitleBar, we don't want any propagation I think).

> Interesting to see incremental vertical shift (rounding when calculating item height?).

Smaller font size/icons = smaller TitleBar = more available height = divided by same item_per_page = slightly different item_heights.
(And I think there is still https://github.com/koreader/koreader/pull/8603#issuecomment-1003901741 which I think previously was centering the content in the available height.)

hius07 commented on frontend/apps/filemanager/filemanager.lua:
```diff
@@ -114,74 +104,26 @@ end
+        right_icon = "plus",
+        right_icon_size_ratio = 1,
+        right_icon_tap_callback = function() self:onShowPlusMenu() end,
+        right_icon_hold_not_propagate = false,
```
Very nice, thank you.

poire-z commented on frontend/ui/widget/titlebar.lua:
```diff
@@ -48,6 +48,7 @@ local TitleBar = OverlapGroup:extend{
     left_icon_size_ratio = 0.6,
     left_icon_rotation_angle = 0,
     left_icon_tap_callback = function() end,
+    -- set hold_callback=false to not handle the event and let it propagate; otherwise the event is discarded
```
May be move this below, after all the icon stuff, as it's valid for any callback (would work with tap too I think):
```diff
    right_icon_hold_callback = function() end,
    right_icon_allow_flash = true,
+        -- set any of these _callback to false to not handle the event
+        -- and let it propagate; otherwise the event is discarded
```

poire-z commented on frontend/apps/filemanager/filemanager.lua:
```diff
@@ -114,74 +104,26 @@ end
+        fullscreen = "true",
+        align = "center",
+        title = self.title,
+        title_top_padding = Screen:scaleBySize(6),
```
Was the `title_top_padding` we does compute if none provided not good ? Or just different from the previous look ? And if different, not good enough ?
The computation also takes into account button_padding that you set to Screen:scaleBySize(5).
How is it if you provide none, or just one button_padding?
Why is `button_padding = Screen:scaleBySize(11), -- fine to keep exit/cross icon diagonally aligned with screen corners ` not good ? Because this was set to look good with `left_icon_size_ratio = 0.6` only, and is bad with `= 1` ?

Just asking, i'm fine with this as-is. I had done enough maths, I don't need more :)

hius07 commented:
Illustrations of TitleBar computations

Icon size 0.6, all defaults
![31](https://user-images.githubusercontent.com/62179190/149131444-d520bd78-251a-458f-a8ff-8b8ad2c7f1f5.png)

Icon size 1.0, all defaults
![32](https://user-images.githubusercontent.com/62179190/149131480-fda2ff40-aa7f-4c6d-915b-8832f77b5a3d.png)

Added `button_padding = Screen:scaleBySize(5)`
![33](https://user-images.githubusercontent.com/62179190/149131528-33bd7d07-9cb6-46b5-abab-e4060c521fe2.png)

Added `title_top_padding = Screen:scaleBySize(6)` (final)
![34](https://user-images.githubusercontent.com/62179190/149131584-b3a247a3-0f69-4bd1-8754-d9143420b753.png)

Frenzie commented on frontend/ui/widget/titlebar.lua:
```diff
@@ -56,6 +56,8 @@ local TitleBar = OverlapGroup:extend{
     right_icon_tap_callback = function() end,
     right_icon_hold_callback = function() end,
     right_icon_allow_flash = true,
+        -- set any of these _callback to false to not handle the event
+        -- and let it propagate; otherwise the event is discarded
```
Indent on purpose?

hius07 replied:
To emphasize that the comment concerns **above** code, I think.

poire-z replied:
> To emphasize that the comment concerns above code, I think.

Yep, that's exactly how I usually use/do it.

Frenzie replied:
In this case a comment at the top would make more sense to me, but /shrug

poire-z replied:
For me, comments above are generic/explaining.
Comments at end of line can be that too, but I may also add some to explain a quick edge case or specific that is not worth being in the generic comment above, as distracting or too edge casy.
So, when that additional edgecasey comment is too long to be put at end of line, I do the trick with indentation below.

Frenzie replied:
I just find it harder to follow. Instead of "here's what this is" it's "here's what this was" which works great for a single line or a line or two but here we're talking about quite a few. :-)
